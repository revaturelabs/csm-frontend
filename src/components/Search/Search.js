import React from "react";
import { Input } from "semantic-ui-react";
import { useSelector, useDispatch } from "react-redux";

const Search = (props) => {
  const dispatch = useDispatch();
  const categories = useSelector((state) => state.swotReducer.categories);
  const findPattern = (event) => {
    let elements = [];
    // let re = /^[^`~!@#$%^&*()_+={}\[\]|\\:;“’<,>.?๐฿]*$/;
    let pattern = event.target.value;
    pattern = pattern.replace(/[.*+\-?^${}()|[\]\\]/g, "\\$&");
    for (const category of categories) {
      if (
        category.toLowerCase().search(pattern.toLowerCase()) > -1
      ) {
        elements.push(category);
      }
    }
    dispatch({
      type: "updateDisplayCategories",
      getDisplayCategories: elements,
    });
  };
  return (
    <Input
      id="search"
      fluid
      onChange={findPattern}
      icon="search"
      placeholder="Search..."
    />
  );
};

export default Search;
