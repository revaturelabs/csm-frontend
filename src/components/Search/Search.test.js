// Component.test.js
import React from "react";
import ReactDOM from "react-dom";
import Enzyme, { shallow, mount, render } from "enzyme";
import EnzymeAdapter from "enzyme-adapter-react-16";
// import { Provider } from "react-redux";
import reducer from "../../reducers/swotReducer";
import state from "../../testImports/swotStoreTest.js";
import Search from "./Search";

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe("Search unit test", () => {
  it("Rendering component without props.", () => {
    const component = shallow(<Search />);
    expect(component).toMatchSnapshot();
  });

  it("Rendering component with props that match the original component.", () => {
    const id = "search";
    const find = jest.fn();
    const icon = "search";
    const placeholder = "Search...";
    const component = shallow(
      <Search
        id={id}
        icon={icon}
        fluid
        onChange={find}
        placeholder={placeholder}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("Mounting, testing user input, and dismounting the component.", () => {
    const id = "search";
    const find = jest.fn();
    const icon = "search";
    const placeholder = "Search...";
    const component = mount(
      <Search
        id={id}
        icon={icon}
        fluid
        onChange={find}
        placeholder={placeholder}
      />
    );
    component.find("#search").value;
    component.simulate("keypress", { keyCode: "A" });
    component.simulate("keypress", { keyCode: "W" });
    component.simulate("keypress", { keyCode: "S" });
    expect(component).toMatchSnapshot();
    component.unmount();
  });

  it("Testing find function call.", () => {
	const { useSelector, useDispatch } = require("react-redux");
    useSelector.mockImplementation((callback) => {
      return callback({
        swotReducer: {
          categories: ["AWS", "Python", "Javascript"],
          displayCategories: [],
        },
      });
	});
	useDispatch.mockImplementation(() => {
		return reducer
	})
    const spyFunc = jest.fn();
    Object.defineProperty(global.document, "querySelector", { value: spyFunc });
    const id = "search";
    const find = jest.fn();
    const icon = "search";
    const placeholder = "Search...";
    const component = mount(
      <Search
        id={id}
        icon={icon}
        fluid
        onChange={find}
        placeholder={placeholder}
      />
    );
    component.find("#search").simulate("keypress", { keyCode: "A" });
	expect(find).toHaveBeenCalled();
	useSelector.mockClear();
  });
});
