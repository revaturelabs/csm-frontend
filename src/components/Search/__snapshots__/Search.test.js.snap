// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search unit test Mounting, testing user input, and dismounting the component. 1`] = `
<Search
  fluid="true"
  icon="search"
  id="search"
  onChange={[MockFunction]}
  placeholder="Search..."
>
  <Input
    fluid={true}
    icon="search"
    id="search"
    onChange={[Function]}
    placeholder="Search..."
    type="text"
  >
    <div
      className="ui fluid icon input"
    >
      <input
        id="search"
        onChange={[Function]}
        placeholder="Search..."
        type="text"
      />
      <Icon
        as="i"
        name="search"
      >
        <i
          aria-hidden="true"
          className="search icon"
          onClick={[Function]}
        />
      </Icon>
    </div>
  </Input>
</Search>
`;

exports[`Search unit test Rendering component with props that match the original component. 1`] = `
<Input
  fluid={true}
  icon="search"
  id="search"
  onChange={[Function]}
  placeholder="Search..."
  type="text"
/>
`;

exports[`Search unit test Rendering component without props. 1`] = `
<Input
  fluid={true}
  icon="search"
  id="search"
  onChange={[Function]}
  placeholder="Search..."
  type="text"
/>
`;
